### PR TITLE
Changes for `devnet-8`

### PR DESCRIPTION
**Pull request by ethDreamer**
## Issue Addressed

* #4443
* #4444
* #4445
* #4487
* #4560

Reviewing this PR is probably easiest by reviewing the individual commits as they are self-contained.

### CODE REVIEW

File: beacon_node/beacon_chain/src/execution_payload.rs
Comment (jimmygchen) [inline beacon_node/beacon_chain/src/execution_payload.rs]: I'm not really sure here, but I _think_ the parent beacon block root would be the `tree_hash_root` of the latest block header from the pre-state? i.e. `state.latest_block_header().canonical_root()`

File: beacon_node/execution_layer/src/engine_api.rs
Comment (jimmygchen) [inline beacon_node/execution_layer/src/engine_api.rs:684]: Is there a reason why this returns an error?

**Comment by jimmygchen** [comment]
I think we may also need to add the `engine_forkchoiceUpdatedV3` method calls to support the new `PayloadAttributes`

https://github.com/sigp/lighthouse/blob/6ef151a6c1d4265ad4dec7467ba141176bb7b705/beacon_node/execution_layer/src/engine_api/http.rs#L1144-L1156

File: beacon_node/execution_layer/src/engine_api.rs
Comment (realbigsean) [inline beacon_node/execution_layer/src/engine_api.rs:425]: We could use the `From<GetPayloadResponseRef> for ExecutionPayloadRef` in these getters to avoid having to do the fork matches

`ExecutionPayloadRef::from(self.to_ref()).fee_recipient()`

File: beacon_node/execution_layer/src/lib.rs
Comment (realbigsean) [inline beacon_node/execution_layer/src/lib.rs]: If we just make this `BUILDER_PROFIT_THRESHOLD` configurable (we should rename it though because we already have a `--builder-profit-threshold` flag) and expose it with a flag, this seems like enough for now. We could also just make it a binary option to listen to the EL suggestion or not

File: consensus/state_processing/src/per_block_processing/verify_attestation.rs
Comment (realbigsean) [inline consensus/state_processing/src/per_block_processing/verify_attestation.rs:49]: Would you mind add a comment here referencing `EIP-7045`

**Review by realbigsean — CHANGES_REQUESTED** [review]
Nice! A few things:

- I think we should bump the EF tests in the PR to get CI working https://github.com/sigp/lighthouse/issues/4519
- I don't see these p2p changes in attestation verification:  https://github.com/ethereum/consensus-specs/pull/3360/files#diff-21cd1dde2a4815e714ef42d0cefa1bbd2997a4358ecf6e9e51025332c1d642fd
  - I think we need to adjust [`verify_propagation_slot_range`](https://github.com/sigp/lighthouse/blob/stable/beacon_node/beacon_chain/src/attestation_verification.rs#L1037) by ditching `ATTESTATION_PROPAGATION_SLOT_RANGE` for deneb
- After implementing 4778, do you have any thoughts about this? I think they are trying to get more CL feedback about whether to merge it. Whatever is easiest, preferable on our end  https://github.com/ethereum/consensus-specs/pull/3454
- We need to update voluntary exit signing in the account manager as well: https://github.com/sigp/lighthouse/blob/stable/account_manager/src/validator/exit.rs

File: beacon_node/execution_layer/src/engine_api.rs
Comment (ethDreamer) [reply beacon_node/execution_layer/src/engine_api.rs:684]: For `deneb`, you can't convert only an `ExecutionPayload` into a `NewPayloadRequest`, you also need the `parent_beacon_block_root`. This conversion is only used for tests thus far. Lighthouse is always using the code above to convert from a `BeaconBlockRef` to a `NewPayloadRequest`.

Comment (realbigsean) [reply beacon_node/execution_layer/src/engine_api.rs:684]: Related to this: https://github.com/ethereum/consensus-specs/pull/3454 
it's a pain point for lodestar I guess, has it impacted our tests?

**Comment by jimmygchen** [comment]
> * We need to update voluntary exit signing in the account manager as well: https://github.com/sigp/lighthouse/blob/stable/account_manager/src/validator/exit.rs

Oh good point, we also need to update voluntary exit signing in the VC API:
https://github.com/sigp/lighthouse/blob/6ef151a6c1d4265ad4dec7467ba141176bb7b705/validator_client/src/validator_store.rs#L663-L670

File: beacon_node/execution_layer/src/lib.rs
Comment (ethDreamer) [reply beacon_node/execution_layer/src/lib.rs]: I'm not sure if this is something we necessarily want to treat in the same way as the `--builder-profit-threshold` setting. It's a difficult question of when do we listen to the execution layer for overriding the builder. I put this trivial mechanism in purely as a safety measure so users won't lose a super valuable block. I think there needs to be more discussion around this involving both EL & CL before we decide how to properly deal with this.

File: beacon_node/execution_layer/src/engine_api.rs
Comment (ethDreamer) [reply beacon_node/execution_layer/src/engine_api.rs:684]: This is just here to make some testing code (which currently doesn't bother to test post-merge forks) more concise. It only impacts some merge fork tests (thus the error condition isn't triggered). There's `TODO`'s in the code to consider making those tests fork-aware / test multiple forks. I we decide to do that, this code will be revisited.

Comment (jimmygchen) [reply beacon_node/execution_layer/src/engine_api.rs:684]: If it's only used in tests, I wonder if it's worth adding `#[cfg(test)]` to make it more explicit and ensure we don't later use this in production code?

Comment (ethDreamer) [reply beacon_node/execution_layer/src/engine_api.rs:684]: Unfortunately we don't have our code structured that way. It's used in the `mock_execution_engine` as well as the execution engine integration tests. I believe the later isn't run with `cfg(test)` enabled. Fixing up these tests is beyond the scope of this PR. I'd be happy to look into doing this but I'm not sure it should be done before we get `deneb-free-blobs` merged into `unstable` as it would only make the review even worse.

Comment (ethDreamer) [reply beacon_node/execution_layer/src/engine_api.rs:684]: As it stands it's not really the worst thing in the world. I would be worried if it were a panic but an error isn't bad.

File: consensus/state_processing/src/per_block_processing/verify_attestation.rs
Comment (ethDreamer) [reply consensus/state_processing/src/per_block_processing/verify_attestation.rs:49]: done!

File: beacon_node/beacon_chain/src/execution_payload.rs
Comment (ethDreamer) [reply beacon_node/beacon_chain/src/execution_payload.rs]: Very nice catch! Didn't realize that was already the parent state. I've updated it now :)

File: beacon_node/execution_layer/src/engine_api.rs
Comment (ethDreamer) [reply beacon_node/execution_layer/src/engine_api.rs:425]: done!

File: beacon_node/execution_layer/src/engine_api.rs
Comment (jimmygchen) [reply beacon_node/execution_layer/src/engine_api.rs:684]: ah yeah that wouldn't work for integration tests, got it thanks :)

File: beacon_node/execution_layer/src/lib.rs
Comment (realbigsean) [reply beacon_node/execution_layer/src/lib.rs]: I think we should just ignore `should_override_builder` for now then, and maybe just log a warning until we have a better idea of how we should interpret it.

Comment (ethDreamer) [reply beacon_node/execution_layer/src/lib.rs]: okay I've updated this to have a new flag:

`ignore-builder-override-suggestion-threshold` which is a percentage difference to be exceeded by the builder payload in order to override this. I gave it a default value of 10%

File: validator_client/src/validator_store.rs
Comment (realbigsean) [inline validator_client/src/validator_store.rs]: do you think it's worth moving this logic to the `self.signing_context()` method? in case it's ever used elsewhere with `Domain::VoluntaryExit`

File: consensus/state_processing/src/per_block_processing/signature_sets.rs
Comment (realbigsean) [inline consensus/state_processing/src/per_block_processing/signature_sets.rs:394]: Similar to the other comment - should we move fork specific logic into `spec.get_domain` so that other potential usages with the voluntary exit domain don't break? Since we pass in fork I think we have enough info

File: beacon_node/execution_layer/src/engine_api/http.rs
Comment (realbigsean) [inline beacon_node/execution_layer/src/engine_api/http.rs]: should this be `forkchoice_updated_v3`?

File: beacon_node/src/cli.rs
Comment (realbigsean) [inline beacon_node/src/cli.rs]: since this parses to an `f32`,  this would also accept decimals right? might be worth making that explict

**Review by realbigsean — COMMENTED** [review]
So the consensus changes for EIP-7045  are very minor but end up creating a decent amount of code duplication.  I'm wondering if we think it's worth making an effort to try to reduce the duplication to make it easier to review on the deneb merge to unstable. If not I think we should probably make it a bit easier to understand what exactly about these methods should be different, maybe by pointing out specifics in method docs. What are your thoughts?

**Comment by ethDreamer** [comment]
> So the consensus changes for EIP-7045 are very minor but end up creating a decent amount of code duplication. I'm wondering if we think it's worth making an effort to try to reduce the duplication to make it easier to review on the deneb merge to unstable.

I was also kind of annoyed by the code duplication, however I think the code is significantly more readable. If attestation processing has changed in `deneb`, I think it would be much more confusing to be calling methods like `compute_beacon_block_attestation_reward_altair`, `new_for_altair`, `altair::process_attestations`, within the `deneb` fork.

As far as removing some duplication, I suppose you *could* make the argument for making `get_attestation_participation_flag_indices` universal (fork-agnostic). But doing this would make the code for `compute_beacon_block_attestation_reward_altair` identical to `compute_beacon_block_attestation_reward_deneb` which is a problem because the fork subscript is required here to differentiate it from `compute_beacon_block_attestation_reward_base`. Thus we would be calling `compute_beacon_block_attestation_reward_altair` from within `deneb` and hiding that it actually behaves differently.

> If not I think we should probably make it a bit easier to understand what exactly about these methods should be different, maybe by pointing out specifics in method docs. What are your thoughts?

I can add comments to the functions if that's what you mean

File: consensus/state_processing/src/per_block_processing/signature_sets.rs
Comment (ethDreamer) [reply consensus/state_processing/src/per_block_processing/signature_sets.rs:394]: Not in this case because [`get_domain`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#get_domain) is a spec-defined function & changes of this type are done via bypassing this function to instead go straight to [`compute_domain`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#compute_domain). Compare `process_voluntary_exit` in:
 * [`phase0`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#voluntary-exits)
  * [`deneb`](https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/beacon-chain.md#modified-process_voluntary_exit)
  
  this puts the change in the spec in approximately the same place as the change in our code and makes our code more spec-conformant overall

File: validator_client/src/validator_store.rs
Comment (ethDreamer) [reply validator_client/src/validator_store.rs]: yes i can do that

File: beacon_node/execution_layer/src/engine_api/http.rs
Comment (ethDreamer) [reply beacon_node/execution_layer/src/engine_api/http.rs]: Yes! Nice catch. I must have gotten it backwards when I wrote it unless it was a typo.

You see once I change this to call `forkchoice_updated_v3`, this will result in lighthouse calling `forkchoice_updated_v3` as soon as the EL says it supports it, which should be **before the `deneb` fork**. Note the [specification](https://github.com/ethereum/execution-apis/blob/main/src/engine/cancun.md#specification-1) says this is an error condition **if we have a `PayloadAttributes`** but says nothing about this when we don't have one. After speaking with some EL devs it turns out to be easier for both of us if we agree that it's **not** an error condition to call `forkchoice_updated_v3` before the `deneb` fork when there is no `PayloadAttributes`.

File: beacon_node/src/cli.rs
Comment (ethDreamer) [reply beacon_node/src/cli.rs]: good idea :)

Comment (ethDreamer) [reply beacon_node/src/cli.rs]: [link to fix](https://github.com/sigp/lighthouse/pull/4518/files#diff-fdb1b9480ecf7dbd64bd5acbb641a040103fb96cc366ad700adf23a4e6aa9c32R1105)

File: beacon_node/execution_layer/src/engine_api/http.rs
Comment (ethDreamer) [reply beacon_node/execution_layer/src/engine_api/http.rs]: [link to fix](https://github.com/sigp/lighthouse/pull/4518/files#diff-8b79597c7e1a33167a01f676e50f09dbc40e761dad5b294086ebda40828e1641R1199)

File: validator_client/src/validator_store.rs
Comment (ethDreamer) [reply validator_client/src/validator_store.rs]: [link to fix](https://github.com/sigp/lighthouse/pull/4518/files#diff-0ce9ffb26be983050e21c205e660fe26a1c34ad451275ae2d5a165f1341cd5cbL373)

File: consensus/state_processing/src/per_block_processing/signature_sets.rs
Comment (realbigsean) [reply consensus/state_processing/src/per_block_processing/signature_sets.rs:394]: Right! Makes sense

**Review by realbigsean — APPROVED** [review]
LGTM

**Comment by realbigsean** [comment]
We're testing some testing networking stuff on devnet 7 still so I think we should wait to merge this for a bit

**Comment by realbigsean** [comment]
There's a couple merge conflicts, but I can make a `devnet-7` branch and we can merge this once that's resolved

**Comment by ethDreamer** [comment]
alright I believe I've gotten rid of the duplicated code :)

**Review by realbigsean — APPROVED** [review]
LGTM

**Pull request merged**
